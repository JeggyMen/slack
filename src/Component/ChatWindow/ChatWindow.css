.chatInfo {
    height: 60px;
    background-color: #8744f3;
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 30px;
    color: aliceblue;
    font-size: 1.5em;
    position: sticky;
    top: 0;
}

.chatIcons i {
    margin-left: 10px; 
    margin-right: 10px; 
    cursor: pointer;
}

.ChatWindow {
    display: flex;
    flex-direction: column;
    height: 100vh; 
    background-color: #f0f0f0; 
}

.messagesContainer {
    flex-grow: 1; 
    overflow-y: auto; 
    padding: 10px;
}

.messages {
    background-color: antiquewhite;
    padding: 10px;
    margin-bottom: 10px;
    display: flex;
    flex-direction: column;
    gap: 10px;
}

.messagesOwn{
    background-color: rgb(224, 158, 70);
    padding: 10px;
    margin-bottom: 10px;
    display: flex;
    flex-direction: column;
    gap: 10px;
}

.input {
    background-color: white;
    padding: 30px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    border: 2px solid rgb(0, 0, 0);
    border-bottom: 3px solid rgb(0, 0, 0);
    border-radius: 2px;
    border-bottom-right-radius: 15px;
    position: sticky;
    bottom: 0;
}

.input > .inputArea {
    width: 100%;
    border: none;
    outline: none;
}

button {
    margin-left: 10px;
}