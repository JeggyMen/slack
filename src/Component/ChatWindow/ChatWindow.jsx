import React, { useState, useEffect } from 'react';
import { Modal, Button } from 'react-bootstrap';
import axios from 'axios';
import { API_URL } from '../Constants/Constants';
import UserServices from '../Services/UserServices';
import './ChatWindow.css';
function ChatWindow(props) {
    const [show, setShow] = useState(false);
    const [text, setText] = useState("");
    const [messages, setMessages] = useState([]);
    const { user, selectedUser } = props;
    const handleClose = () => setShow(false);
    const handleShow = () => setShow(true);
    const fetchMessages = async () => {
        if (selectedUser) {
            const fetchedMessages = await UserServices.fetchMessages(user, selectedUser);
            setMessages(fetchedMessages);
        }
    };
    useEffect(() => {
        fetchMessages();
    }, [selectedUser]);
    const handleSendMessage = async () => {
        if (!text.trim()) {
            return;
        }
        const messageData = {
            receiver_id: selectedUser.id,
            receiver_class: 'User',
            body: text,
        };
        const headers = {
            'access-token': user.accessToken,
            'client': user.client,
            'expiry': user.expiry,
            'uid': user.uid,
        };
        try {
            const response = await axios.post(`${API_URL}/messages`, messageData, { headers });
            if (response.status === 200) {
                const newMessage = response.data.data;
                setMessages(prevMessages => [...prevMessages, newMessage]);
                setText('');
            }
        } catch (error) {
            if (error.response && error.response.status === 401) {
                console.error('Unauthorized: Invalid or expired token');
            } else {
                console.error('Failed to send message:', error);
            }
        }
    };
    const getUserNameFromEmail = (email) => {
        if (!email) return "";
        const userName = email.split('@')[0];
        return userName.split('.')[0];
    };
    return (
        <div className='ChatWindow'>
            <div className="chatInfo">
                <span>{selectedUser?.email}</span>
                <div className="chatIcons">
                    <i className="bi bi-camera-video-fill"></i>
                    <i className="bi bi-person-add"></i>
                    <i className="bi bi-three-dots" onClick={handleShow}></i>
                </div>
            </div>
            <div className='messagesContainer'>
                {messages && messages.map((message, index) => (
                    <div key={index} className={message.sender?.email === user.email ? 'messagesOwn' : 'messages'}>
                        <div className='texts'>
                            <p><strong>{getUserNameFromEmail(message.sender?.email)}:</strong> {message.body}</p>
                            <span>{new Date(message.created_at).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })}</span>
                        </div>
                    </div>
                ))}
            </div>
            <div className='input'>
                <input
                    type="text"
                    className="inputArea"
                    placeholder='Type something...'
                    value={text}
                    onChange={e => setText(e.target.value)}
                />
                <button onClick={handleSendMessage}>Send</button>
            </div>
            <Modal show={show} onHide={handleClose}>
                <Modal.Header closeButton>
                    <Modal.Title>Chat Details</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <div>
                        <h5>Chat with</h5>
                        <p>{selectedUser?.email}</p>
                    </div>
                </Modal.Body>
                <Modal.Footer>
                    <Button variant="secondary" onClick={handleClose}>
                        Close
                    </Button>
                </Modal.Footer>
            </Modal>
        </div>
    );
}
export default ChatWindow;