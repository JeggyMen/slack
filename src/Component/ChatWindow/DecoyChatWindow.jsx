import React from 'react';
import { Modal, Button } from 'react-bootstrap';
import './ChatWindow.css';


function DecoyChatWindow(props) {
    const { show, setShow } = props;
    const handleClose = () => setShow(false);
    const handleShow = () => setShow(true);
    return (
        <div className='ChatWindow'>
            <div className="chatInfo">
                <span>No user selected</span>
                <div className="chatIcons">
                    <i className="bi bi-camera-video-fill"></i>
                    <i className="bi bi-person-add"></i>
                    <i className="bi bi-three-dots" onClick={handleShow}></i>
                </div>
            </div>
            <div className='messagesContainer'>
                <div className='noMessages'>
                    <p>Please select a user to start chatting.</p>
                </div>
            </div>
            <div className='input'>
                <input
                    type="text"
                    className="inputArea"
                    placeholder='Type something...'
                    disabled
                />
                <button disabled>Send</button>
            </div>
            <Modal show={show} onHide={handleClose}>
                <Modal.Header closeButton>
                    <Modal.Title>Chat Details</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <div>
                        <h5>No user selected</h5>
                        <p>Please select a user to view chat details.</p>
                    </div>
                </Modal.Body>
                <Modal.Footer>
                    <Button variant="secondary" onClick={handleClose}>
                        Close
                    </Button>
                </Modal.Footer>
            </Modal>
        </div>
    );
}
export default DecoyChatWindow;