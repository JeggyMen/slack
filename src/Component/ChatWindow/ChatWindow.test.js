import { render, screen, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import ChatWindow from "./ChatWindow";

test("render the ChatWindowqq component", () => {
    render(<ChatWindow />);
});

test("snapshot test for ChatWindow", () => {
    const counter = renderer.create(<ChatWindow />).toJSON();
    expect(counter).toMatchSnapshot();
});
