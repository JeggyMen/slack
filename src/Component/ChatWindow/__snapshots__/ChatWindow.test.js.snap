// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for ChatWindow 1`] = `
<div
  className="ChatWindow"
>
  <div
    className="chatInfo"
  >
    <span />
    <div
      className="chatIcons"
    >
      <i
        className="bi bi-camera-video-fill"
      />
      <i
        className="bi bi-person-add"
      />
      <i
        className="bi bi-three-dots"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="messagesContainer"
  />
  <div
    className="input"
  >
    <input
      className="inputArea"
      onChange={[Function]}
      placeholder="Type something..."
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Send
    </button>
  </div>
</div>
`;
